client/src/components/signup/SignupForm.js

```diff
@@ -0,0 +1,128 @@
+  const [errors, setErrors] = useState({
+    username: '',
+    email: '',
+    password: '',
+  })
+
+  // isFormValid function is a function that runs only in case of the "signup" option.
+  // This functions checks all the fields and updates the error messages accordingly.
+  // This function returns true if all the fields are vaid and false when at least one
+  // field is invalid.
+  const isFormValid = () => {
+    // Create temporary variables to store the validation status of each field
+    const fieldValidationErrors = {}
+    const isValidUsername = inputValues.username.length >= 5
+    const isValidEmail = (/^([\w.%+-]+)@([\w-]+\.)+([\w]{2,})$/i).test(inputValues.email)
+    const isValidPassword = inputValues.password.length >= 5
+
+    // Based on the above validation assign the error message, or remove the error message
+    // in the fieldValidationErrors variable.
+    fieldValidationErrors.username = isValidUsername ? '' : 'minimum 5 characters'
+    fieldValidationErrors.email = isValidEmail ? '' : 'invalid email address'
+    fieldValidationErrors.password = isValidPassword ? '' : 'minimum 5 characters'
+
+    // Assign the saved errors (fieldValidationErrors) to the errors state variable.
+    setErrors({ ...fieldValidationErrors })
+
+    // If all the vields passed the validation return true, otherwise return false.
+    if (isValidUsername && isValidEmail && isValidPassword) { return true }
+    return false
+  }
+
+  // Create handleInput function runs on every change of any input field.
+  const handleInput = (e) => {
+    // Store name (as property) and value of the current input field.
+    const property = e.target.name
+    const { value } = e.target
+
+    // Update inputValues variable. It will update the current value of the used field.
+    setInputValues({
+      ...inputValues,
+      [property]: value,
+    })
+  }
+
+  // handleSubmit function activates onSubmit when the submit button is pressed.
+  const handleSubmit = (e) => {
+    e.preventDefault()
+    // Check if we are in the Signup mode and the signup form is valid
+    // OR
+    // If we are in the Login mode
+    if ((formChoice === 'signup' && isFormValid()) || (formChoice === 'login')) {
+      // Run the onValidateSubmit function which is the function passed from the
+      // parent component as a prop. This function delivers the validated credentials.
+      onValidatedSubmit({ ...inputValues })
+    }
+  }
+
+  return (
+    <div className={styles.SignupForm__container}>
+      <h4 className={styles.SignupForm__heading}>
+        {formChoice === 'login' && 'Login To Your Account'}
+        {formChoice === 'signup' && 'Create Your Account'}
+      </h4>
+      <form
+        onSubmit={(event) => handleSubmit(event)}
+        className={styles.SignupForm}
+      >
+        <SignupFormField
+          type="text"
+          name="username"
+          placeholder="Your Name"
+          iconSrc={userIcon}
+          handleInput={handleInput}
```

Comment:
You are even  using the recommended implementation here. You pointed to `handleInput` without passing the `event` object. So, it will work fine for the `handleSubmit`.

Reply:
True. I thought before that when passing function through components you don't need to include the event object but you have to include it at the final component that will actually be calling that function.